ddl/check.lua

```diff
@@ -628,75 +628,141 @@ local function check_index(i, index, space)
+            space.name, 'bucket_id', #space.indexes['bucket_id'].parts
+        )
+    end
+
+    if space.indexes['bucket_id'].parts[1].path ~= 'bucket_id' then
+        return nil, string.format(
+            "spaces[%q].indexes[%q].parts[1].path: invalid field reference " ..
+            "(reference to bucket_id expected, got %s)",
+            space.name, 'bucket_id', space.indexes['bucket_id'].parts[1].path
+        )
+    end
+
+    return true
+end
+
+local function is_callable(object)
+    if type(object) == 'function' then
         return true
     end
 
-    do -- check that format['bucket_id'] valid
-        if not space.fields.bucket_id then
-            return nil, string.format(
-                "spaces[%q].format: sharding_key exists in the space, but there's" ..
-                " no bucket_id defined in 'format' section",
-                space.name
-            )
-        end
+    if type(object) == 'cdata' then
+        return true
+    end
 
-        if space.fields.bucket_id.type ~= 'unsigned' then
-            return nil, string.format(
-                "spaces[%q].format[%q].type: bad value (unsigned expected, got %s)",
-                space.name, 'bucket_id', space.fields.bucket_id.type
-            )
-        end
+    if type(object) == 'userdata' then
+        return true
+    end
+
+    local object_metatable = getmetatable(object)
+    if type(object) == 'table' and object_metatable
+    and object_metatable.__call then
+        return type(object_metatable.__call) == 'function'
+    end
+
+    return false
+end
+
+local function check_custom_sharding_func(sharding_func_name)
+    local sharding_func = rawget(_G, sharding_func_name)
+    if sharding_func == nil then
+        return false
     end
 
-    do -- check that indexes['bucket_id'] valid
-        if not space.indexes.bucket_id then
+    return is_callable(sharding_func)
+end
+
+local function check_sharding_func(space)
+    if not space.sharding_func then
+        return true
+    end
+
+    do -- check that sharding_func format is valid
+        if type(space.sharding_func) ~= 'string' and type(space.sharding_func) ~= 'table' then
             return nil, string.format(
-                "spaces[%q].indexes: sharding_key exists in the space, but there's" ..
-                " no bucket_id defined in 'indexes' section",
-                space.name
+                "spaces[%q].sharding_func: bad value" ..
+                " (string or table expected, got %s)",
+                space.name, type(space.sharding_func)
             )
         end
 
-        if space.indexes.bucket_id.unique then
-            return nil, string.format(
-                "spaces[%q].indexes[%q].unique: bucket_id index can't be unique",
-                space.name, 'bucket_id'
-            )
+        if type(space.sharding_func) == 'string' then
+            if not space.sharding_func:match('^[%w_]+[%.[%w_]+%.]*[%w_]+$')
```

Comment:
Comment in code added. This check is for dot.notation: sharding func name should not begin with dot or digit and last symbol may not be dot. If sharding func name  has at least one dot according regex then it is a dot.notation and we don't check this name in `_G` and checking of possibility of a call  happens in #80 , otherwise this is a name of user sharding func that should be checked in `_G`.